vss_collection.py

```diff
@@ -0,0 +1,183 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+from typing import List
+
+from vspec.model.vsstree import VSSNode
+
+from utils import CodeGeneratorContext
+
+_COLLECTION_SUFFIX = "Collection"
+_COLLECTION_REG_EX = r"\w+\[\d+,(\d+)\]"
+
+_TYPE_SUFFIX = "Type"
+_DEFAULT_RANGE_NAME = "element"
+
+"""VSS Collection helper."""
+
+
+class VssInstance:
+    """VSS Instance Model."""
+
+    def __init__(self, name: str, content: List[str], is_range: bool):
+        """Construct of new insntance object."""
+        # Content List
+        self.content = content
+
+        # The name of the Instance
+        self.name = name
+
+        # Is a named range flag that will be used to generate
+        # getter function for element by index.
+        self.is_range = is_range
+
+
+class VssCollection:
+    """VSS Collection Object."""
+
+    def __init__(self, node: VSSNode, ctx: CodeGeneratorContext):
+        """Construct of new collection object."""
+        self.ctx = ctx
+        self.name = f"{node.name}{_COLLECTION_SUFFIX}"
+        self.content_list = self.__gen_collection(node)
+
+    def __parse_instances(self, reg_ex, instance) -> VssInstance:
+        result = list()
+
+        # parse string instantiation elements (e.g. Row[1,5])
+        range_name = _DEFAULT_RANGE_NAME
+        if isinstance(instance, str):
+            if re.match(reg_ex, instance):
+                inst_range_arr = re.split(r"\[+|,+|\]", instance)
+                range_name = inst_range_arr[0]
+                lower_bound = int(inst_range_arr[1])
+                upper_bound = int(inst_range_arr[2])
+                for element in range(lower_bound, upper_bound + 1):
+                    result.append(f"{range_name}{element}")
+
+                return VssInstance(range_name, result, True)
+
+            raise ValueError("", "", f"instantiation type {instance} not supported")
+
+        # Use list elements for instances (e.g. ["LEFT","RIGHT"])
+        if isinstance(instance, list):
+            for element in instance:
+                result.append(f"{element}")
+            return VssInstance(range_name, result, False)
+
+        raise ValueError("", "", f"is of type {type(instance)} which is unsupported")
+
+    def __gen_collection_types(self, name, instances) -> tuple[str, list]:
+        _type_name = f"{name}{_TYPE_SUFFIX}"
+        result = list()
+        print(f"{' ' * 5}- {_type_name:25}{instances}")
+        result.append(f"{self.ctx.tab}class {_type_name}(Model):")
+        result.append(f"{self.ctx.tab * 2}def __init__(self, name, parent):")
+        result.append(f"{self.ctx.tab * 3}super().__init__(parent)")
+        result.append(f"{self.ctx.tab * 3}self.name = name")
```

Comment:
please use `with ctx as new_context` instead of self.ctx.tab everywhere. or where not possible use indent()